pkg/apis/scheduling/v1alpha1/types_placement.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/apis/conditions/v1alpha1"
+)
+
+// placement defines a selection rule to choose ONE location for MULTIPLE namespaces in a workspace.
+//
+// placement is in Pending state initially, when a location is selected by the placement, the placement
+// turns to UnBound state. In Pending or UnBound state, the selection rule can be updated to select another location.
+// When the user annotates a namespace with the key of "scheudling.kcp.dev/placement", the namespace will
+// pick one placement, and this placement is transfered to Bound state. The spec of the placement is immutable
+// in Bound state. The placement will turns back to Unbound state when no namespace uses this
+// placement any more.
+//
+// +crd
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope=Cluster,categories=kcp
+type Placement struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec PlacementSpec `json:"spec,omitempty"`
+
+	// +optional
+	Status PlacementStatus `json:"status,omitempty"`
+}
+
+type PlacementSpec struct {
+	// loacationSelectors represents a slice of label selector to select a location, these label selectors
+	// are logically ORed.
+	LoacationSelectors []metav1.LabelSelector `json:"locationSelectors,omitempty"`
+
+	// locationResource is the group-version-resource of the instances that are subject to the locations to select.
+	//
+	// +required
+	// +kubebuilder:Required
+	LocationResource GroupVersionResource `json:"locationResource"`
+
+	// namespaceSelector is a label selector to select ns. It match all ns by default, but can be specified to
+	// a certain set of ns.
+	NamespaceSelector metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+
+	// locationWorkspace is an absolute reference to a workspace for the location. If it is not set, the workspace of
+	// APIBinding will be used.
+	// +optional
+	// +kubebuilder:validation:Pattern:="^root(:[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$"
+	LocationWorkspace *string `json:"locationWorkspace,omitempty"`
```

Comment:
do we have any requirements here? In other context, we require this to be a sibling, an ancestor or sibling of ancestor.

Reply:
Reason: such a rule forces publicly consumed workspaces to be as high in the hierarchy as required by its consumers.

Reply:
oh, we actually have that:
> The workspace must be some ancestor or a child of some ancestor.
:+1:

Reply:
nit: omitempty string does not have to be a pointer to be optional.